File: frontend/src/services/email_template.ts
import { apiService } from './api';
import type {
  EmailTemplateCreate,
  EmailTemplateUpdate,
  EmailTemplateResponse,
  EmailTemplatePreview,
  EmailTemplatePreviewResponse,
  EmailTaskCreate,
  EmailTaskUpdate,
  EmailTaskResponse,
  EmailSendRequest,
  EmailSendResponse,
  EmailTemplateListParams
} from '@/models/email_template';

export class EmailTemplateService {
  private readonly baseUrl = '/email-templates';

  // Email Template CRUD 操作
  async createTemplate(data: EmailTemplateCreate): Promise<EmailTemplateResponse> {
    return apiService.post<EmailTemplateResponse>(this.baseUrl, data);
  }

  async listTemplates(params?: EmailTemplateListParams): Promise<EmailTemplateResponse[]> {
    return apiService.get<EmailTemplateResponse[]>(this.baseUrl, params);
  }

  async getTemplate(id: number): Promise<EmailTemplateResponse> {
    return apiService.get<EmailTemplateResponse>(`${this.baseUrl}/${id}`);
  }

  async updateTemplate(id: number, data: EmailTemplateUpdate): Promise<EmailTemplateResponse> {
    return apiService.put<EmailTemplateResponse>(`${this.baseUrl}/${id}`, data);
  }

  async deleteTemplate(id: number): Promise<void> {
    return apiService.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Email Template 預覽
  async previewTemplate(previewData: EmailTemplatePreview): Promise<EmailTemplatePreviewResponse> {
    return apiService.post<EmailTemplatePreviewResponse>(
      `${this.baseUrl}/preview`,
      previewData
    );
  }

  // 立即發送 Email
  async sendEmail(data: EmailSendRequest): Promise<EmailSendResponse> {
    return apiService.post<EmailSendResponse>(`${this.baseUrl}/send`, data);
  }

}

// Email Task 相關服務（使用現有的 scheduler API）
export class EmailTaskService {
  private readonly baseUrl = '/scheduler';

  async createEmailTask(data: EmailTaskCreate): Promise<EmailTaskResponse> {
    // 轉換為 ScheduledTaskCreate 格式
    const taskData = {
      name: data.name,
      description: data.description,
      schedule_expression: data.schedule_expression,
      timezone: data.timezone,
      target_type: 'email',
      target_arn: data.recipients[0], // 使用第一個收件人作為 target_arn
      target_input: {
        use_template: data.use_template,
        template_id: data.template_id,
        template_variables: data.template_variables,
        subject: data.subject,
        body: data.body,
        html_body: data.html_body,
        recipients: data.recipients,
        cc: data.cc,
        bcc: data.bcc,
        sender: data.sender
      }
    };

    return apiService.post<EmailTaskResponse>(this.baseUrl, taskData);
  }

  async updateEmailTask(id: number, data: EmailTaskUpdate): Promise<EmailTaskResponse> {
    // 轉換為 ScheduledTaskUpdate 格式
    const taskData = {
      name: data.name,
      description: data.description,
      schedule_expression: data.schedule_expression,
      timezone: data.timezone,
      target_input: {
        use_template: data.use_template,
        template_id: data.template_id,
        template_variables: data.template_variables,
        subject: data.subject,
        body: data.body,
        html_body: data.html_body,
        recipients: data.recipients,
        cc: data.cc,
        bcc: data.bcc,
        sender: data.sender,
      },
    };

    return apiService.put<EmailTaskResponse>(`${this.baseUrl}/${id}`, taskData);
  }

  async getEmailTasks(): Promise<EmailTaskResponse[]> {
    return apiService.get<EmailTaskResponse[]>(`${this.baseUrl}?target_type=email`);
  }

  async getEmailTask(id: number): Promise<EmailTaskResponse> {
    return apiService.get<EmailTaskResponse>(`${this.baseUrl}/${id}`);
  }

  async deleteEmailTask(id: number): Promise<void> {
    return apiService.delete<void>(`${this.baseUrl}/${id}`);
  }

  async executeEmailTask(id: number): Promise<void> {
    return apiService.post<void>(`${this.baseUrl}/${id}/execute`);
  }

  async updateEmailTaskState(id: number, state: string): Promise<void> {
    return apiService.put<void>(`${this.baseUrl}/${id}/state`, { state });
  }
}

export const emailTemplateService = new EmailTemplateService();
export const emailTaskService = new EmailTaskService();

File: src/routers/email_template.py
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import List, Optional
from src.models.pydantic.email_template import (
    EmailTemplateCreate,
    EmailTemplateUpdate,
    EmailTemplateResponse,
    EmailTemplatePreview,
    EmailTemplatePreviewResponse
)
from src.services.email_template import EmailTemplateService
from src.dependencies.services import get_email_template_service
from src.utils.logger import logger

router = APIRouter()


@router.get("", response_model=List[EmailTemplateResponse])
async def list_templates(
        is_active: Optional[bool] = Query(None, description="是否啟用"),
        search: Optional[str] = Query(None, description="搜索關鍵字"),
        limit: int = Query(100, ge=1, le=1000, description="返回數量限制"),
        offset: int = Query(0, ge=0, description="偏移量"),
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """獲取 Email 模板列表"""
    return await service.list_templates(
        is_active=is_active,
        search=search,
        limit=limit,
        offset=offset
    )


@router.post("", response_model=EmailTemplateResponse)
async def create_email_template(
        template_data: EmailTemplateCreate,
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """創建新的 Email 模板"""
    try:
        return await service.create_template(template_data)
    except Exception as e:
        logger.error(f"創建模板失敗: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{template_id}", response_model=EmailTemplateResponse)
async def get_email_template(
        template_id: int,
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """獲取單個 Email 模板"""
    try:
        template = await service.get_template(template_id)
        if not template:
            raise HTTPException(status_code=404, detail="模板不存在")
        return template
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"獲取模板失敗: {str(e)}")
        raise HTTPException(status_code=500, detail="獲取模板失敗")


@router.put("/{template_id}", response_model=EmailTemplateResponse)
async def update_email_template(
        template_id: int,
        template_data: EmailTemplateUpdate,
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """更新 Email 模板"""
    try:
        template = await service.update_template(template_id, template_data)
        if not template:
            raise HTTPException(status_code=404, detail="模板不存在")
        return template
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新模板失敗: {str(e)}")
        raise HTTPException(status_code=500, detail="更新模板失敗")


@router.delete("/{template_id}", status_code=204)
async def delete_email_template(
        template_id: int,
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """刪除 Email 模板"""
    try:
        await service.delete_template(template_id)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"刪除模板失敗: {str(e)}")
        raise HTTPException(status_code=500, detail="刪除模板失敗")


@router.post("/preview", response_model=EmailTemplatePreviewResponse)
async def preview_email_template(
        preview_data: EmailTemplatePreview,
        service: EmailTemplateService = Depends(get_email_template_service)
):
    """預覽 Email 模板"""
    try:
        return await service.preview_template(preview_data)
    except Exception as e:
        logger.error(f"預覽模板失敗: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))
